File: Runtime/Playgap/Plugins/iOS/PlaygapAdsBridge.swift
import Playgap

@objc
public final class PlaygapAdsBridge: NSObject {

    private static var showHandler: ShowHandler?
    private static var claimRewardsHandler: ClaimRewardsHandler?

    @objc
    public static func observeNetwork(observer: @escaping (Bool) -> Void) {
        PlaygapAds.networkObserver = observer
    }

    @objc
    public static func initialize(apiKey: String, completion: @escaping (Error?) -> Void) {
        PlaygapAds.initialize(apiKey: apiKey, completion: completion)
    }

    @objc
    public static func showRewarded(
        onShowFailed: @escaping (_ error: String) -> Void,
        onShowImpression: @escaping (_ impressionId: String) -> Void,
        onShowPlaybackEvent: @escaping (_ period: String) -> Void,
        onShowCompleted: @escaping () -> Void,
        onUserEarnedReward: @escaping (_ rewardId: String) -> Void
    ) {
        let handler = ShowHandler(
            showFailed: onShowFailed,
            showImpression: onShowImpression,
            showPlaybackEvent: onShowPlaybackEvent,
            showCompleted: onShowCompleted,
            userEarnedReward: onUserEarnedReward
        )

        PlaygapAds.showRewarded(from: UnityGetGLViewController(), delegate: handler)
        showHandler = handler
    }

    @objc
    public static func showInterstitial(
        onShowFailed: @escaping (_ error: String) -> Void,
        onShowImpression: @escaping (_ impressionId: String) -> Void,
        onShowPlaybackEvent: @escaping (_ period: String) -> Void,
        onShowCompleted: @escaping () -> Void,
        onUserEarnedReward: @escaping (_ rewardId: String) -> Void
    ) {
        let handler = ShowHandler(
            showFailed: onShowFailed,
            showImpression: onShowImpression,
            showPlaybackEvent: onShowPlaybackEvent,
            showCompleted: onShowCompleted,
            userEarnedReward: onUserEarnedReward
        )

        PlaygapAds.showInterstitial(from: UnityGetGLViewController(), delegate: handler)
        showHandler = handler
    }

    @objc
    public static func claimRewards(
        onRewardScreenShown: @escaping () -> Void,
        onRewardScreenFailed: @escaping (_ error: String) -> Void,
        onRewardScreenClosed: @escaping () -> Void,
        onStoreClick: @escaping () -> Void,
        onUserClaimedRewards: @escaping (_ rewardIds: [String]) -> Void
    ) {
        let handler = ClaimRewardsHandler(
            rewardScreenShown: onRewardScreenShown,
            rewardScreenFailed: onRewardScreenFailed,
            rewardScreenClosed: onRewardScreenClosed,
            storeClick: onStoreClick,
            userClaimedRewards: onUserClaimedRewards
        )

        PlaygapAds.claimRewards(from: UnityGetGLViewController(), delegate: handler)
        claimRewardsHandler = handler
    }

    @objc
    public static func checkRewards() -> String? {
        guard let rewards = PlaygapAds.checkRewards() else { return nil }

        let mapped = RewardsObjc(unclaimed: rewards.unclaimed, claimed: rewards.claimed)

        guard let json = try? JSONEncoder().encode(mapped) else { return nil }

        return String(data: json, encoding: String.Encoding.utf8)
    }

    @objc
    public static func sendEvent(
        _ eventType: String,
        _ payload: String,
        _ objectId: String
    ) {
        PlaygapAnalytics.sendEvent(eventType, adId: objectId, payloadJson: payload)
    }
}

private final class ShowHandler: ShowDelegate {

    private let showFailed: (_ error: String) -> Void
    private let showImpression: (_ impressionId: String) -> Void
    private let showPlaybackEvent: (_ period: String) -> Void
    private let showCompleted: () -> Void
    private let userEarnedReward: (_ rewardId: String) -> Void

    init(
        showFailed: @escaping (_: String) -> Void,
        showImpression: @escaping (_: String) -> Void,
        showPlaybackEvent: @escaping (_ period: String) -> Void,
        showCompleted: @escaping () -> Void,
        userEarnedReward: @escaping (_: String) -> Void
    ) {
        self.showFailed = showFailed
        self.showImpression = showImpression
        self.showPlaybackEvent = showPlaybackEvent
        self.showCompleted = showCompleted
        self.userEarnedReward = userEarnedReward
    }

    func onShowFailed(error: Playgap.PlaygapAds.ShowError) {
        showFailed(error.localizedDescription)
    }

    func onShowImpression(impressionId: String) {
        showImpression(impressionId)
    }

    func onUserEarnedReward(_ reward: Playgap.Reward) {
        userEarnedReward(reward.id)
    }

    func onShowPlaybackEvent(_ event: PlaybackEvent) {
        showPlaybackEvent(event.rawValue)
    }

    func onShowCompleted() {
        showCompleted()
    }
}

private final class ClaimRewardsHandler: ClaimRewardsDelegate {

    private let rewardScreenShown: () -> Void
    private let rewardScreenFailed: (_ error: String) -> Void
    private let rewardScreenClosed: () -> Void
    private let storeClick: () -> Void
    private let userClaimedRewards: (_ rewardIds: [String]) -> Void
    private var claimedOfflineRewards: [String] = []

    init(
        rewardScreenShown: @escaping () -> Void,
        rewardScreenFailed: @escaping (_: String) -> Void,
        rewardScreenClosed: @escaping () -> Void,
        storeClick: @escaping () -> Void,
        userClaimedRewards: @escaping (_: [String]) -> Void
    ) {
        self.rewardScreenShown = rewardScreenShown
        self.rewardScreenFailed = rewardScreenFailed
        self.rewardScreenClosed = rewardScreenClosed
        self.storeClick = storeClick
        self.userClaimedRewards = userClaimedRewards
    }

    public func onRewardScreenShown() {
        rewardScreenShown()
    }

    public func onRewardScreenFailed(_ error: Playgap.PlaygapAds.ClaimRewardsError) {
        rewardScreenFailed(error.localizedDescription)
    }

    public func onRewardScreenClosed() {
        if !claimedOfflineRewards.isEmpty {
            userClaimedRewards(claimedOfflineRewards)
            claimedOfflineRewards.removeAll()
        }

        rewardScreenClosed()
    }

    public func onStoreClick() {
        storeClick()
    }

    public func onUserClaimedReward(_ reward: Playgap.Reward) {
        claimedOfflineRewards.append(reward.id)
    }
}

struct RewardsObjc: Encodable {

    let unclaimed: [String]
    let claimed: [String]
}

File: Runtime/Playgap/Plugins/Android/PlaygapAdsBridge.java
package io.playgap.sdk;

import android.app.Activity;
import android.content.Context;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.util.ArrayList;
import java.util.List;
import java.util.HashMap;
import java.lang.reflect.Type;

class PlaygapAdsBridge {

    private static final List<String> claimedOfflineRewards = new ArrayList<String>();

    public static void registerNetworkObserver(Context context, NetworkListenerBridge listener) {
        PlaygapAds.Companion.registerNetworkObserver(context, listener::onNetworkStateChanged);
    }
    
    public static void initialize(Context context, String apiKey, InitializationListenerBridge listener) {
        PlaygapAds.Companion.initialize(context, apiKey, new InitializationListener() {

            @Override
            public void onInitializationError(InitError error) {
                listener.onInitializationError(error.getType().description(context));
            }

            @Override
            public void onInitialized() {
                listener.onInitialized();
            }
        });
    }

    public static void showRewarded(Activity activity, ShowListenerBridge listener) {
        PlaygapAds.Companion.showRewarded(activity, new ShowListener() {
            @Override
            public void onUserEarnedReward(PlaygapReward playgapReward) {
                listener.onUserEarnedReward(playgapReward.getId());
            }

            @Override
            public void onShowCompleted() {
                listener.onShowCompleted();
            }

            @Override
            public void onShowPlaybackEvent(PlaybackEvent event) {
                listener.onShowPlaybackEvent(event.raw());
            }

            @Override
            public void onShowImpression(String impressionId) {
                listener.onShowImpression(impressionId);
            }

            @Override
            public void onShowFailed(ShowError error) {
                listener.onShowFailed(error.getType().description(activity));
            }
        });
    }

    public static void showInterstitial(Activity activity, ShowListenerBridge listener) {
        PlaygapAds.Companion.showInterstitial(activity, new ShowListener() {
            @Override
            public void onUserEarnedReward(PlaygapReward playgapReward) {
                listener.onUserEarnedReward(playgapReward.getId());
            }

            @Override
            public void onShowCompleted() {
                listener.onShowCompleted();
            }

            @Override
            public void onShowPlaybackEvent(PlaybackEvent event) {
                listener.onShowPlaybackEvent(event.raw());
            }

            @Override
            public void onShowImpression(String impressionId) {
                listener.onShowImpression(impressionId);
            }

            @Override
            public void onShowFailed(ShowError error) {
                listener.onShowFailed(error.getType().description(activity));
            }
        });
    }

    public static String checkRewards() {
        PlaygapRewards rewards = PlaygapAds.Companion.checkRewards();
        if (rewards == null) {
            return null;
        }

        return new Gson().toJson(rewards);
    }

    public static void claimRewards(Activity activity, ClaimRewardsListenerBridge listener) {
        PlaygapAds.Companion.claimRewards(activity, new ClaimRewardsListener() {
            
            @Override
            public void onRewardScreenShown() {
                listener.onRewardScreenShown();
            }

            @Override
            public void onRewardScreenFailed(ClaimRewardError error) {
                listener.onRewardScreenFailed(error.getType().description(activity));
            }

            @Override
            public void onRewardScreenClosed() {
                if (!claimedOfflineRewards.isEmpty()) {
                    listener.onUserClaimedRewards(claimedOfflineRewards.toArray(new String[0]));
                    claimedOfflineRewards.clear();
                }

                listener.onRewardScreenClosed();
            }

            @Override
            public void onUserClaimedReward(PlaygapReward reward) {
                claimedOfflineRewards.add(reward.getId());
            }
        });
    }

    public static void sendEvent(String eventType, String payload, String objectId) {
        Gson gson = new Gson();

        Type mapType = new TypeToken<HashMap<String, String>>(){}.getType();
        HashMap<String, String> payloadMap = gson.fromJson(payload, mapType);
        
        PlaygapAnalytics.Companion.sendEvent(eventType, payloadMap, objectId);
    }
}

interface NetworkListenerBridge {
    void onNetworkStateChanged(Boolean isConnected);
}

interface InitializationListenerBridge {
    void onInitializationError(String error);

    void onInitialized();
}

interface ShowListenerBridge {
    void onShowFailed(String error);

    void onShowImpression(String impressionId);

    void onShowPlaybackEvent(String event);

    void onShowCompleted();

    void onUserEarnedReward(String rewardId);
}

interface ClaimRewardsListenerBridge {

    void onRewardScreenShown();

    void onRewardScreenFailed(String error);

    void onRewardScreenClosed();

    void onUserClaimedRewards(String[] rewardIds);
}